[ONNX] Create GraphContext and load `g.op` method to the class

justinchuby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85488
* #85487
* #85486
* #85350
* #85349
* #85312
* #85165
* #85491
* #84776
* __->__ #84728

This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`
- Create `torchscript.py` to house torch.Graph related functions
- Change `GraphContext.op` to create nodes in the Block instead of the Graph
- Create `add_op_with_blocks` to handle scenarios where we need to directly manipulate sub-blocks. Update loop and if symbolic functions to use this function.

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

### Decision

Keep GraphContext flatand note that it will change in the future.

justinchuby committed:
[ONNX] Graph context

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

AllenTiTaiWang commented on torch/onnx/symbolic_helper.py:
```diff
@@ -459,7 +459,7 @@ def _scalar(x: Any) -> Optional[Number]:
 
 
 @_beartype.beartype
-def _if_scalar_type_as(g: _C.Graph, self, tensor):
```
It's unused?

justinchuby replied:
Yes

AllenTiTaiWang commented on torch/onnx/utils.py:
```diff
@@ -1774,32 +1781,39 @@ def _run_symbolic_function(
-                    ctx = _exporter_states.SymbolicContext(_params_dict, env, n, block)
-                    return symbolic_fn(ctx, g, *inputs, **attrs)
+                    ctx = _exporter_states.SymbolicContext(
+                        _params_dict, env, node, block, opset_version
```
I know it's not from this PR, but `_params_dict` seems a dangerous global variable here. I wonder how we can make it safer. I also see your previous PR tag it as TODO: remove its usage from other cpp file.

justinchuby replied:
It's ancient code. We should look to remove it from future PRs.

AllenTiTaiWang commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,423 @@
+    def inputs(self) -> List[_C.Value]:
+        ...
+
+    def outputs(self) -> List[_C.Value]:
```
I wonder if these are needed for creating this object.

thiagocrepaldi replied:
IIUC, this is just an Interface class which implementation is calling the implementor method with same name. It mimics `torch.Graph`

I agree it is a bit confusing because the docstring mentions that it "implements" all methods from `torch.Graph`, and then it just declares them.

Maybe the docstring could be reworked to let explicit the intent of serving as abstract class

justinchuby replied:
Updated the docstring

AllenTiTaiWang commented:
I really like this PR, as it's making converting process clearer than ever. I remember the time I was struggling to find how g.op() works and so on. 

>Should we put all the context inside SymbolicContext and make it an attribute in the GraphContext class? This way we only >define two attributes GraphContext.graph and GraphContext.context. Currently all context attributes are directly defined in >the class.

Does this affect user to use `SymbolicContext` independently? I see `SymbolicContext` was introduced as an important parameter for custom op registration. If it doesn't, I think it's a good idea.

titaiwangms reviewed (approved):
LGTM!

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

thiagocrepaldi commented:
This is nice.

As mentioned at [#84776 review](https://github.com/pytorch/pytorch/pull/84776#pullrequestreview-1113965911), passing context object for *all* symbolics might be useful to improve JIT Script during ONNX export. Hopefully we can use this approach to resolve `Tensor[]` into something more specific that includes `dtype` and `shape`

Let's discuss this approach a little further to make sure `GraphContext` would be accessible at `torch.jit.script` time or in other parts JIT script process the model. Would be possible to come up with a couple slides or a one pager to discuss on a meeting? Maybe our colleagues have more insight on how this could be amplified for a higher impact

thiagocrepaldi commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,426 @@
+
+
+@runtime_checkable
+class GraphLike(Protocol):
```
Duplicating each  method defined at https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python/python_ir.cpp#L218 is error prone. Users may tap new methods to the pybind class and not change this one (because they dont know about it).

Did you consider creating a `GraphAbstract` class using `class GraphInterface(metaclass=abc.ABCMeta)` to define an Interface and then change https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python/python_ir.cpp#L218 to implement such interface?

With that, `GraphAbstract` is the only source of truth for the public interface and the `Graph` and/or `GraphContext` implement them as needed, inheriting from the same class

justinchuby replied:
I am conscious about creating dependencies on Python code in cpp. They are usually harder to debug. This Protocol is for type checkers and ides only. If the cpp code is changed, we can expect the new method usable because getattr in GraphContext looks for any method that is not defined in class.

justinchuby replied:
I am going to remove `GraphLike` and use GraphContext for everything

BowenBao replied:
I wonder what's the life cycle design for `GraphContext`. Is it for symbolic function use only, hence created every time in `_run_symbolic_function`, or is it something that will be passed around though out entire export? If it's the latter, we should take cpp interactions into consideration.

justinchuby replied:
For now it is created every time in `_run_symbolic_function` so there is no interactions yet. Any ideas?

thiagocrepaldi commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,426 @@
+
+
+@runtime_checkable
+class GraphLike(Protocol):
```
```suggestion
class _GraphLike(Protocol):
```
If we don't want users to instantiate it, better to prepend a `_` to make it private?

justinchuby replied:
GraphLike is used in the symbolic function annotations, so it is exported outside of this module (public in that sense). Since torchscript is in _internal, the module itself is private and we don't give users any guarantee. In the future when we want to export some of the features to users we can come back and decide if parts of the module needs refactoring.

thiagocrepaldi commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,426 @@
+
+    # Relay methods from _C.Graph for compatibility with symbolic functions that expect
+    # a _C.Graph
+    def __getattr__(self, name: str) -> Any:
```
@AllenTiTaiWang This is where those methods are "implemented"

thiagocrepaldi commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,426 @@
+    The set of operators and the inputs/attributes they take
+    is documented at https://github.com/onnx/onnx/blob/master/docs/Operators.md
+
+    This function is monkey-patched onto Graph.
```
Why monkey-patching `Graph` and not defining it there?

In code documentation, "why" might be more important than the "what" or "how", given what/how is described by the source-code.

justinchuby replied:
This function is moved from _patch_torch unchanged. I will add a comment of why

BowenBao replied:
What is the process of getting rid of monkey patching _C.Graph, if that's intended?

justinchuby replied:
I think it's the meta internal tests I am concern about. Do you think this will affect them?

thiagocrepaldi commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,426 @@
+"""Utilities for manipulating the torch.Graph object and the torchscript."""
```
`torchscript.py` is too generic. Torch Script can refer to the TorchScript "language", `torch.jit.script` engine or maybe TorchScript IR.

`_scalar` and `_is_caffe2_aten_fallback` don't depend/implement on anything related to TorchScript in any of the context above, for example. Instead, they process `torch.Tensor` properties or globals defined at `torch.onnx` namespace

If we want to mix torch script and helpers in a single file, `utils.py` still looks a good name. Or we split files to keep different concepts decoupled

In fact, aren't some of the helpers below not related to torch script? Examples would be `class GraphLike` which is used by

justinchuby replied:
> `torchscript.py` is too generic. Torch Script can refer to the TorchScript "language", `torch.jit.script` engine or maybe TorchScript IR.

I agree. Any naming suggestions?

> `_scalar` and `_is_caffe2_aten_fallback` don't depend/implement on anything related to TorchScript in any of the context above, for example. Instead, they process `torch.Tensor` properties or globals defined at `torch.onnx` namespace

These are private helper functions that we need to implement the public methods.

thiagocrepaldi reviewed (commented):
Would love to see if we can leverage `GraphContext` when `ScriptModule` is the model being exported

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

BowenBao reviewed (commented):
Got a silly question, why is it a bad idea to inherit Graph, and add `op` and `at` methods, as well as other attributes to it?

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby commented:
> Got a silly question, why is it a bad idea to inherit Graph, and add `op` and `at` methods, as well as other attributes to it?

I think that actually works

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby commented:
We need to change what g.op returns. If Tuple is returned mypy will scream. But if we don't include Tuple beartype will scream. 😢

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`
- Move monkey patches on torch Graph and block etc to `torchscript.py`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`
- Move monkey patches on torch Graph and block etc to `torchscript.py`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby commented on torch/onnx/_patch_torch.py:
```diff
@@ -15,7 +15,16 @@
 
-# TODO(#78694): Refactor the patching process to make it more transparent to users.
+# TODO(#78694): Remove this file after PyTorch 1.14.
+# Add functions in this file are deprecated and should not be used
```
```suggestion
# All functions in this file are deprecated and should not be used
```

justinchuby commented on torch/onnx/_patch_torch.py:
```diff
@@ -100,6 +115,12 @@ def _aten_op(g: _C.Graph, operator: str, *args, overload_name: str = "", **kwarg
+@_deprecation.deprecated(
+    "1.13",
+    "1.14",
+    "note 'b.op()' is to be removed from torch.Graph. Please open a"
```
```suggestion
    "note 'b.op()' is to be removed from torch.Block. Please open a"
```

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`
- Move monkey patches on torch Graph and block etc to `torchscript.py`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`
- Move monkey patches on torch Graph and block etc to `torchscript.py`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`
- Move monkey patches on torch Graph and block etc to `torchscript.py`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby commented:
@BowenBao PTAL

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`
- Move monkey patches on torch Graph and block etc to `torchscript.py`

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`
- Create `torchscript.py` to house torch.Graph related functions
- Change `GraphContext.op` to create nodes in the Block instead of the Graph
- Create `add_op_with_blocks` to handle scenarios where we need to directly manipulate sub-blocks. Update loop and if symbolic functions to use this function.

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

### Decision

Keep GraphContext flatand note that it will change in the future.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,334 @@
+    for _ in range(n_blocks):
+        new_block = node.addBlock()
+        # Create new copy of the graph context and update the block
+        new_context = GraphContext(
```
dataclasses.replace?

justinchuby replied:
Thanks, learned something new!

BowenBao commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,334 @@
+    new_contexts = []
+    for _ in range(n_blocks):
+        new_block = node.addBlock()
+        # Create new copy of the graph context and update the block
```
perhaps explicitly mention shallow copy?

justinchuby replied:
Done

BowenBao reviewed (approved):
Overall LGTM, a few inline comments.
Please sync with Thiago regarding file name before merging.

BowenBao commented:
> We need to change what g.op returns. If Tuple is returned mypy will scream. But if we don't include Tuple beartype will scream. 😢

Is this something wrong with our code or is it a mypy problem?

justinchuby commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,334 @@
+        outputs: int = 1,
+        **kwargs,
+    ):
+        """Creates an ONNX operator "opname", taking "args" as inputs and attributes "kwargs".
```
Done

justinchuby commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,334 @@
+        The set of operators and the inputs/attributes they take
+        is documented at https://github.com/onnx/onnx/blob/master/docs/Operators.md
+
+        This function is monkey-patched onto Graph.
```
Good catch thanks

justinchuby commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,334 @@
+            outputs: The number of outputs this operator returns.
+                By default an operator is assumed to return a single output.
+                If `outputs` is greater than one, this functions returns a tuple
+                of output `Node`, representing each output of the ONNX operator
```
Done

justinchuby commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,334 @@
+                that takes a list of integers).
+
+        Returns:
+            The node representing the single output of this operator (see the `outputs`
```
Done

justinchuby commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,334 @@
+        outputs: The number of outputs this operator returns.
+            By default an operator is assumed to return a single output.
+            If `outputs` is greater than one, this functions returns a tuple
+            of output `Node`, representing each output of the ONNX operator
```
Done

justinchuby commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,334 @@
+
+    Returns:
+        (Union[_C.Value, Tuple[_C.Value, ...]])
+        The node representing the single output of this operator (see the `outputs`
```
Done

justinchuby commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,334 @@
+    n_blocks: int = 1,
+    **attributes,
+) -> Tuple[Any, Tuple[GraphContext, ...], _C.Node]:
+    """Creates an ONNX operator "opname", taking "args" as inputs and attributes "kwargs".
```
Updated

justinchuby commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,334 @@
+
+    Returns:
+        A tuple of (output_values, new_contexts, node) where:
+            output_values: The node representing the single output of this operator
```
rephrased

justinchuby commented on torch/onnx/_internal/torchscript.py:
```diff
@@ -0,0 +1,334 @@
+        outputs: The number of outputs this operator returns.
+            By default an operator is assumed to return a single output.
+            If `outputs` is greater than one, this functions returns a tuple
+            of output `Node`, representing each output of the ONNX operator
```
Done

justinchuby commented:
> Is this something wrong with our code or is it a mypy problem?

If we annotate the return type to be a union of Value and a tuple of Values, every time we use the output as either a Value or a Tuple, we will need to assert its type first, otherwise mypy will say certain methods are not implemented or the object, because of course methods in a tuple are not in a Value and vice versa. I haven't found a nice way of doing it. One possibility is to split the multi output case to a new function which always returns a tuple.

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`
- Create `torchscript.py` to house torch.Graph related functions
- Change `GraphContext.op` to create nodes in the Block instead of the Graph
- Create `add_op_with_blocks` to handle scenarios where we need to directly manipulate sub-blocks. Update loop and if symbolic functions to use this function.

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

### Decision

Keep GraphContext flatand note that it will change in the future.

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Create GraphContext and load `g.op` method to the class"


This PR create the `GraphContext` class and relays all graph methods to _C.Graph as well as implements the `g.op`  method. The GraphContext object is passed into the symbolic functions in place of _C.Graph for compatibility with existing symbolic functions.

This way (1) we can type annotate all `g` args because the method is defined and (2) we can use additional context information in symbolic functions. (3) no more monkey patching on `_C.Graph`

Also

- Fix return type of `_jit_pass_fixup_onnx_controlflow_node`
- Create `torchscript.py` to house torch.Graph related functions
- Change `GraphContext.op` to create nodes in the Block instead of the Graph
- Create `add_op_with_blocks` to handle scenarios where we need to directly manipulate sub-blocks. Update loop and if symbolic functions to use this function.

## Discussion

Should we put all the context inside `SymbolicContext` and make it an attribute in the `GraphContext` class? This way we only define two attributes `GraphContext.graph` and `GraphContext.context`. Currently all context attributes are directly defined in the class.

### Decision

Keep GraphContext flatand note that it will change in the future.

[ghstack-poisoned]

justinchuby commented:
@pytorchbot merge -g
